### DISCUSSION TITLE
Streamlining CI/Ops for visual and baseline tests

### DISCUSSION THREAD

**Discussion by bryevdv** [bokeh/bokeh#11893] [Build / CI]
Roughly transcribing from a tangential PR comment thread in 

* https://github.com/bokeh/bokeh/pull/11856

The discussion seemed to get 👍 from four people: @jbednar @p-himik @ianthomas23 and of course myself. Opening a proper discussion here to make sure the thread is not lost. 

---

I think we should consider picking a chrome version for each release branch, and sticking with that one version the entire length of the branch. I.e. I would propose

* A new `branch-x.y` is created
* At the same time, a new `bokeh-dev-x.y` image is also published
* This image is both the "user dev" image as well as the image that runs the test in CI
* If there is some exceptional need, we can rev to `bokeh-dev-x.y-1`, etc 

I am proposing a tradeoff here, namely that we stop running the baseline tests on OSX and Windows, and we stop updating to follow chrome all the time, in order to get our ongoing ops/maintenance burden down to manageable levels. I am really, truly OK with this tradeoff. I can't point to a single issue that was detected ***only***  due to baseline differences that affect OSX or Windows, but did not affect Linux. But I can point to an *enormous* amount of time expended on painful ops churn just for its own sake. 

This is an explicit recognition of ***the reality of our limited organizational capacity***, and the fact that we are struggling heavily under the current system. I have worked for large organizations that have tens of people working full-time to support testing/ops on projects honestly not much bigger than bokeh (and definitely with less downloads). We simply ***do not have that level of resourcing***, and need to adjust accordingly. The value provided by the amount of baseline testing we currently do is not justified by its cost to our much smaller team. If we adopt the above scheme, then:

* Anyone can generate a new baseline and reference images completely locally, and not have to go through a multi-CI-run process, downloading artifacts from arcane GH pages and unzipping them in precise locations. This is almost worth the change all by itself
* We will only have one image to maintain at a time (per active release branch). Users using the docker workflow will be guaranteed to be in sync with CI, and ready to go without more work from us.
* Docker image can also potential be an option of last resort for folks who can't get webdrivers working 
* Exploding baseline and image storage and bandwidth costs will be reduced. 

cc @bokeh/dev

**Comment by bryevdv** [comment]
In the PR, @ianthomas23 replied

> I like the idea of dropping the osx and windows baseline images and only using the linux ones. I think this will be particularly helpful in getting new contributors through their first PR.
> 
> For naming of the version-specific images, it would be more docker-like to use bokeh-dev:x.y rather than bokeh-dev-x.y, i.e. keeping the same image name (before the colon) but changing the tag (after the colon). But this is just a detail, the system is flexible enough to support either way of working or indeed to change it in future if we have a better/different idea.
> 
> I agree that using a docker image to run the tests in CI should be pretty easy.
> 
> One feature I have added is setting an environment variable BOKEH_IN_DOCKER=1 whenever the image is run in a container, so it may be possible to keep a single set of scripts for testing, etc, that can be used both inside and outside of an image and just have additions based on whether the env var is set or not. This might be better than duplicating scripts.
> 
>

**Reply by bryevdv** [reply]
I think `bokeh-dev:x.y` is perfectly fine

**Comment by mattpap** [comment]
Although the workflow around using docker for testing is not yet fully established, perhaps we could drop windows and osx baselines now anyway? That would greatly simplify work on layout and reduce repo churn.

**Reply by bryevdv** [reply]
I support that 100%
